# 04 – Details

Esplorazione e data cleaning del dataset `details.csv`.

**Colonne:**
| Colonna | Descrizione |
|---|---|
| `mal_id` | ID univoco dell'anime su MyAnimeList |
| `title` | Titolo dell'anime |
| `title_japanese` | Titolo in giapponese |
| `url` | URL della pagina MAL dell'anime |
| `image_url` | URL dell'immagine di copertina |
| `type` | Tipo (TV, Movie, etc) |
| `status` | Stato di trasmissione |
| `score` | Punteggio medio |
| `scored_by` | Numero di utenti che hanno votato |
| `start_date` | Data di inizio trasmissione |
| `end_date` | Data di fine trasmissione |
| `synopsis` | Sinossi testuale |
| `rank` | Classifica globale |
| `popularity` | Popolarità |
| `members` | Numero di utenti che hanno l'anime nella loro lista |
| `favorites` | Numero di utenti che hanno messo l'anime tra i preferiti |
| `genres` | Lista di generi |
| `studios` | Lista di studi di produzione |
| `themes` | Lista di temi narrativi |
| `demographics` | Lista di target demografici |
| `source` | Materiale di origine (Manga, Novel, Original, etc) |
| `rating` | Classificazione per età |
| `episodes` | Numero di episodi |
| `season` | Stagione di uscita |
| `year` | Anno di uscita |
| `producers` | Lista di produttori |
| `explicit_genres` | Lista di generi espliciti |
| `licensors` | Lista di licenziatari |
| `streaming` | Lista di piattaforme di streaming |

## 1. Import e caricamento dati
Importiamo le librerie necessarie e carichiamo il file csv. Facciamo una esplorazione generica per capire la struttura e le caratteristiche del dataset.

In [50]:
import pandas as pd
import numpy as np
from dataset_analyzer import analyze

df_details = pd.read_csv('../datasets/details.csv')
print(f'Shape: {df_details.shape}')
df_details.info()
df_details.head()

Shape: (28955, 29)
<class 'pandas.DataFrame'>
RangeIndex: 28955 entries, 0 to 28954
Data columns (total 29 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   mal_id           28955 non-null  int64  
 1   title            28955 non-null  str    
 2   title_japanese   28836 non-null  str    
 3   url              28955 non-null  str    
 4   image_url        28955 non-null  str    
 5   type             28888 non-null  str    
 6   status           28955 non-null  str    
 7   score            18882 non-null  float64
 8   scored_by        18882 non-null  float64
 9   start_date       28104 non-null  str    
 10  end_date         11267 non-null  str    
 11  synopsis         23828 non-null  str    
 12  rank             21997 non-null  float64
 13  popularity       28955 non-null  int64  
 14  members          28955 non-null  int64  
 15  favorites        28955 non-null  int64  
 16  genres           28955 non-null  str    
 17  stud

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,demographics,source,rating,episodes,season,year,producers,explicit_genres,licensors,streaming
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,NaN,2010-01-01T00:00:00+00:00,...,[],Original,G - All Ages,1.0,NaN,NaN,['Nagoya Zokei University'],[],[],[]
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503.0,2023-06-11T00:00:00+00:00,...,[],Original,PG-13 - Teens 13 or older,1.0,NaN,NaN,[],[],[],[]
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745.0,2007-01-18T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,"['Bandai Visual', 'CyberConnect2']",[],[],[]
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373.0,2007-12-22T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],"['Funimation', 'Bandai Entertainment']",[]
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317.0,2008-03-25T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],[],[]


Il dataset contiene **28.955 righe** e **29 colonne**. Alcune colonne presentano valori nulli. Alcuni dtype richiedono conversione: le date sono in formato stringa, e alcune colonne intere (`scored_by`, `rank`, `episodes`, `year`) sono memorizzate come `float64`. Quiesti veranno gestiti singolarmente nella sezione 2. 

## 1.1 Rimozione duplicati esatti

Prima dell'analisi per colonna, rimuoviamo le righe con valori identici in **tutte** le colonne, mantenendo solo la prima occorrenza.

In [ ]:
n_originale = len(df_details)

mask_dup = df_details.duplicated(keep=False)       # tutte le occorrenze coinvolte
n_righe_coinvolte = mask_dup.sum()                  # righe totali che hanno almeno un duplicato
n_gruppi = df_details[mask_dup].duplicated(keep='first').sum()  # occorrenze extra (da rimuovere)
n_tenute = n_righe_coinvolte - n_gruppi             # prime occorrenze mantenute

print(f'Righe totali coinvolte in duplicazioni : {n_righe_coinvolte:,}')
print(f'  → prime occorrenze mantenute         : {n_tenute:,}')
print(f'  → occorrenze extra rimosse           : {n_gruppi:,}')
print()

df_details.drop_duplicates(keep='first', inplace=True)
print(f'Righe prima della rimozione : {n_originale:,}')
print(f'Righe dopo la rimozione     : {len(df_details):,}')

Righe totali coinvolte in duplicazioni : 0
  → prime occorrenze mantenute         : 0
  → occorrenze extra rimosse           : 0

Righe prima della rimozione : 28,955
Righe dopo la rimozione     : 28,955


## 2. Analisi colonna per colonna

### 2.1 `mal_id`

È la **chiave primaria** del dataset. Deve essere intera, non nulla e univoca. Utilizziamo `analyze` per verificare queste proprietà.

In [51]:
analyze(df_details['mal_id'])


  Nome serie                     mal_id
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   28,955  (100.00%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            62,590
  Range                          62,589
  Media                          33,977.52
  Mediana                        37,969.00
  Moda/e                         1, 5, 6
  Dev. standard    

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un ID univoco.
- **Nessun duplicato**: ogni riga corrisponde a un anime distinto.
- **Dtype corretto**: `int64`, nessuna conversione necessaria.

Nessuna pulizia necessaria.

### 2.2 `title`

Colonna testuale con il titolo dell'anime. Verifichiamo la presenza di valori nulli e duplicati.

In [52]:
analyze(df_details['title'])


  Nome serie                     title
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,953  (99.99%)
  Valori duplicati               2

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  186  caratteri
  Lunghezza media                24.70  caratteri
  Lunghezza mediana              21.0000  caratteri
  Dev. standard lunghezza        15.07
  IQR lunghezza                  18.0000

Valori di lunghezza estrema
───────────────────

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un titolo.
- **Quasi univoco**: 28.953 valori distinti su 28.955. Ci solo 2 titoli duplicati che verifichiamo successivamente. 
- **Lunghezze variabili**: da 1 carattere (titoli come `2`, `9`, singole lettere) a 186 caratteri (lunghi titoli isekai). Vacciamo delle verifiche per approfondire. 
- **14 titoli solo numerici** (es. `2`, `9`). Indaghiamo in seguito.
- **1 titolo contiene un URL** (~0.0%). Ingaghiamo in seguito.
- **43.4% contiene caratteri speciali**: normale, i titoli includono punteggiatura, simboli e trattini.

In [ ]:
# Titoli duplicati 
dup_mask = df_details.duplicated(subset=['title'], keep=False)
print(f'Titoli duplicati (righe coinvolte): {dup_mask.sum()}')
print()
print(df_details[dup_mask][['mal_id', 'title', 'type', 'status', 'year']].to_string(index=False))

# Titoli solo numerici 
print()
print('─' * 60)
num_mask = df_details['title'].str.fullmatch(r'\d+')
print(f'Titoli solo numerici: {num_mask.sum()}')
print()
print(df_details[num_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Titolo che contiene un URL 
print()
print('─' * 60)
url_mask = df_details['title'].str.contains(r'https?://', regex=True, na=False)
print(f'Titoli con URL: {url_mask.sum()}')
print()
print(df_details[url_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Lunghezze estreme 
print()
print('─' * 60)
print('Titoli più corti (len ≤ 2):')
short_mask = df_details['title'].str.len() <= 2
print(df_details[short_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))
print()
print('Titoli più lunghi (len ≥ 100):')
long_mask = df_details['title'].str.len() >= 100
print(df_details[long_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

Titoli duplicati (righe coinvolte): 4

 mal_id                                    title  type          status  year
  59994                                Moonlight Music Finished Airing  <NA>
  60524                                Moonlight Music Finished Airing  <NA>
  53367 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing  <NA>
  55740 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing  <NA>

────────────────────────────────────────────────────────────
Titoli solo numerici: 14

 mal_id  title  type  year
  29978    001   ONA  <NA>
  60145   1135 Music  <NA>
  37065    123 Music  <NA>
  55204   1985 Music  <NA>
  30234   1989 Music  <NA>
  38756      2 Music  <NA>
  54066   2005 Music  <NA>
  32813   2010 Music  <NA>
  53789     32 Music  <NA>
  53771   3341 Music  <NA>
  60486    520 Movie  <NA>
  13971 663114 Movie  <NA>
  41457     86    TV  2021
  36089      9 Music  <NA>

────────────────────────────────────────────────────────────
Titoli con URL: 1

 m

**Risultati dell'indagine:**

- **Titoli duplicati (2 coppie, 4 righe):** `Moonlight` e `Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian` compaiono ciascuno due volte, ma con `mal_id` distinti. Sono anime separati su MAL che condividono il titolo. **Nessuna rimozione necessaria.**

- **Titoli solo numerici (14 righe):** tutti legittimi. La maggior parte sono video musicali (`Music`) o cortometraggi in cui titoli brevi o numerici sono comuni. **Nessuna azione necessaria.**

- **Titolo contenente un URL (1 riga):** `Himitsukessha Taka no Tsume The Movie 3: http://takanotsume.jp wa Eien ni` (`mal_id` 8357). L'URL è parte integrante del titolo ufficiale del film, caratteristica della serie *Taka no Tsume* nota per titoli volutamente assurdi. **Nessuna azione necessaria.**

- **Lunghezze estreme:** i titoli molto corti (1–2 caratteri) sono anime reali con titoli brevi; i titoli molto lunghi (fino a 186 caratteri) sono tipici degli produzioni moderni. Nessun valore anomalo.

**Conclusione:** la colonna `title` non richiede alcuna pulizia. Tutti i casi apparentemente anomali sono valori legittimi.

### 2.3 `title_japanese`

Colonna testuale con il titolo in giapponese. I null sono ammessi: anime di produzione non giapponese o senza titolo kanji registrato su MAL non avranno questo campo.

In [9]:
analyze(df_details['title_japanese'])


  Nome serie                     title_japanese
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,836  (99.59%)
  Null / NaN                     119  (0.41%)
  Stringhe vuote                 0
  Valori unici                   27,772  (95.91%)
  Valori duplicati               1,064

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  85  caratteri
  Lunghezza media                12.24  caratteri
  Lunghezza mediana              10.0000  caratteri
  Dev. standard lunghezza        8.69
  IQR lunghezza                  10.0000

Valori di lunghezza estrema
───────

**Osservazioni:**
- **119 valori nulli** (~0.4%): strutturali, anime di produzione non giapponese o senza titolo kanji registrato su MAL.

Nessuna pulizia necessaria.

In [10]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in title_japanese : {df_details["title_japanese"].isna().sum()} ({df_details["title_japanese"].isna().mean()*100:.1f}%)')

Null in title_japanese : 119 (0.4%)


### 2.4 `url`

URL della pagina MAL dell'anime. Essendo derivato direttamente dal `mal_id`, ci aspettiamo assenza di null e unicità.

In [11]:
analyze(df_details['url'])


  Nome serie                     url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,955  (100.00%)
  Valori duplicati               0

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  35  caratteri
  Lunghezza max                  216  caratteri
  Lunghezza media                60.19  caratteri
  Lunghezza mediana              57.0000  caratteri
  Dev. standard lunghezza        14.92
  IQR lunghezza                  17.0000

Valori di lunghezza estrema
───────────────────

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un URL valido.
- **Nessun duplicato**: coerente con `mal_id` come chiave primaria.

Nessuna pulizia necessaria.

In [12]:
# Nessuna operazione richiesta
print(f'Null in url      : {df_details["url"].isna().sum()}')
print(f'Duplicati in url : {df_details["url"].duplicated().sum()}')

Null in url      : 0
Duplicati in url : 0


### 2.5 `image_url`

URL dell'immagine di copertina. MAL usa un'immagine placeholder per gli anime privi di copertina reale; la identifichiamo e aggiungiamo una colonna booleana `has_image` per distinguere i due casi.

In [13]:
DEFAULT_IMG = 'https://cdn.myanimelist.net/img/sp/icon/apple-touch-icon-256.png'
placeholder = (df_details['image_url'] == DEFAULT_IMG).sum()
print(f'Immagini placeholder : {placeholder:,} ({placeholder/len(df_details)*100:.1f}%)')
print(f'Immagini reali       : {len(df_details)-placeholder:,} ({(len(df_details)-placeholder)/len(df_details)*100:.1f}%)')
print()
analyze(df_details['image_url'])

Immagini placeholder : 198 (0.7%)
Immagini reali       : 28,757 (99.3%)


  Nome serie                     image_url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,758  (99.32%)
  Valori duplicati               197

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  50  caratteri
  Lunghezza max                  64  caratteri
  Lunghezza media                54.69  caratteri
  Lunghezza mediana              56.0000  caratteri
  Dev. standard lunghezza        1.93
  IQR lungh

**Osservazioni:**
- **198 URL placeholder** (~0.7%): corrispondono all'immagine generica di MAL, segnale di assenza di copertina reale.
- **Aggiunta colonna `has_image`**: booleano che distingue gli anime con copertina reale (`True`) da quelli con placeholder (`False`).

In [14]:
DEFAULT_IMG = 'https://cdn.myanimelist.net/img/sp/icon/apple-touch-icon-256.png'
df_details['has_image'] = df_details['image_url'] != DEFAULT_IMG
print(f'has_image → True (reale): {df_details["has_image"].sum():,}  |  False (placeholder): {(~df_details["has_image"]).sum():,}')

has_image → True (reale): 28,757  |  False (placeholder): 198


### 2.6 `type`

Colonna categorica con il tipo di anime (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special). I null sono attesi per gli anime `Not yet aired` il cui tipo non è ancora registrato su MAL.

In [15]:
analyze(df_details['type'])


  Nome serie                     type
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,888  (99.77%)
  Null / NaN                     67  (0.23%)
  Stringhe vuote                 0
  Valori unici                   9  (0.03%)
  Valori duplicati               28,879

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  10  caratteri
  Lunghezza media                3.73  caratteri
  Lunghezza mediana              3.0000  caratteri
  Dev. standard lunghezza        1.81
  IQR lunghezza                  3.0000

Valori di lunghezza estrema
──────────────────────────

**Osservazioni:**
- **67 valori nulli**: strutturali, anime con stato `Not yet aired` per cui il tipo non è ancora registrato su MAL.
- **9 valori distinti**: tutti coerenti con la tassonomia MAL (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special).

Nessuna pulizia necessaria.

In [16]:
# Nessuna operazione richiesta: i null sono strutturali (Not yet aired)
print(f'Null in type  : {df_details["type"].isna().sum()}')
print(f'Valori unici  : {sorted(df_details["type"].dropna().unique().tolist())}')

Null in type  : 67
Valori unici  : ['CM', 'Movie', 'Music', 'ONA', 'OVA', 'PV', 'Special', 'TV', 'TV Special']


### 2.7 `status`

Colonna categorica con lo stato di trasmissione. I valori attesi sono `Finished Airing`, `Currently Airing` e `Not yet aired`.

In [17]:
analyze(df_details['status'])


  Nome serie                     status
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   3  (0.01%)
  Valori duplicati               28,952

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  13  caratteri
  Lunghezza max                  16  caratteri
  Lunghezza media                14.97  caratteri
  Lunghezza mediana              15.0000  caratteri
  Dev. standard lunghezza        0.29
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
─────────────────────

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno uno stato definito.
- **3 valori distinti**: `Finished Airing`, `Currently Airing`, `Not yet aired`, coerenti con la tassonomia MAL.

Nessuna pulizia necessaria.

In [18]:
# Nessuna operazione richiesta
print(f'Null in status : {df_details["status"].isna().sum()}')
print(df_details['status'].value_counts())

Null in status : 0
status
Finished Airing     28097
Not yet aired         544
Currently Airing      314
Name: count, dtype: int64


### 2.8 `score` e `scored_by`

Le due colonne sono co-nulle: quando un anime non ha voti sufficienti su MAL, entrambe sono `NaN`.

In [19]:
analyze(df_details['score'])


  Nome serie                     score
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   561  (1.94%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.8900
  Max                            9.2900
  Range                          7.4000
  Media                          6.3905
  Mediana                        6.3600
  Moda/e                         6.2900
  Dev. standard      

In [20]:
analyze(df_details['scored_by'])


  Nome serie                     scored_by
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   9,245  (31.93%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            101.0000
  Max                            2,979,733
  Range                          2,979,632
  Media                          29,963.08
  Mediana                        1,528.00
  Moda/e                         120.0000

**Osservazioni:**
- **10.073 valori nulli** (~34.8%) in entrambe le colonne: strutturali, anime senza voti sufficienti per avere uno score su MAL. Co-nullità confermata.
- **`score`**: dtype `float64` corretto (valori decimali). Nessuna conversione necessaria.
- **`scored_by`**: dtype `float64` → convertito a `Int64` (nullable integer), dato che i valori sono interi.

In [21]:
# Verifica co-nullità
print(f'Null in score     : {df_details["score"].isna().sum():,}')
print(f'Null in scored_by : {df_details["scored_by"].isna().sum():,}')
print(f'Co-nulli (entrambi NaN) : {(df_details["score"].isna() & df_details["scored_by"].isna()).sum():,}')
print()

df_details['scored_by'] = df_details['scored_by'].astype('Int64')
print(f'scored_by dtype : {df_details["scored_by"].dtype}')

Null in score     : 10,073
Null in scored_by : 10,073
Co-nulli (entrambi NaN) : 10,073

scored_by dtype : Int64


### 2.9 `start_date`

Data di inizio trasmissione, memorizzata come stringa ISO 8601 con timezone UTC. Va convertita in `datetime64[us, UTC]`. I null sono attesi per gli anime `Not yet aired` senza data confermata.

In [22]:
analyze(df_details['start_date'])


  Nome serie                     start_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,104  (97.06%)
  Null / NaN                     851  (2.94%)
  Stringhe vuote                 0
  Valori unici                   9,367  (32.35%)
  Valori duplicati               18,737

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
───────────

**Osservazioni:**
- **851 valori nulli** (~2.9%): strutturali, anime `Not yet aired` senza data confermata.
- **Conversione dtype**: da stringa ISO 8601 a `datetime64[us, UTC]`.

In [23]:
df_details['start_date'] = pd.to_datetime(df_details['start_date'], errors='coerce', utc=True)
print(f'start_date dtype : {df_details["start_date"].dtype}')
print(f'Null residui     : {df_details["start_date"].isna().sum()}')
print(f'Range            : {df_details["start_date"].min()} → {df_details["start_date"].max()}')

start_date dtype : datetime64[us, UTC]
Null residui     : 851
Range            : 1917-01-01 00:00:00+00:00 → 2027-01-01 00:00:00+00:00


### 2.10 `end_date`

Data di fine trasmissione, stessa struttura di `start_date`. I null sono strutturalmente più numerosi: gli anime `Currently Airing` e `Not yet aired` non hanno ancora una data di fine.

In [24]:
analyze(df_details['end_date'])


  Nome serie                     end_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               11,267  (38.91%)
  Null / NaN                     17,688  (61.09%)
  Stringhe vuote                 0
  Valori unici                   5,607  (19.36%)
  Valori duplicati               5,660

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
──────────

**Osservazioni:**
- **17.688 valori nulli** (~61.1%): strutturali, anime `Currently Airing` o `Not yet aired` non hanno ancora una data di fine.
- **Conversione dtype**: da stringa ISO 8601 a `datetime64[us, UTC]`.

In [25]:
df_details['end_date'] = pd.to_datetime(df_details['end_date'], errors='coerce', utc=True)
print(f'end_date dtype : {df_details["end_date"].dtype}')
print(f'Null residui   : {df_details["end_date"].isna().sum()}')
print(f'Range          : {df_details["end_date"].min()} → {df_details["end_date"].max()}')

end_date dtype : datetime64[us, UTC]
Null residui   : 17688
Range          : 1962-02-25 00:00:00+00:00 → 2026-04-10 00:00:00+00:00


### 2.11 `synopsis`

Testo della sinossi dell'anime. I null sono strutturali: cortometraggi, CM e PV spesso non hanno una sinossi registrata su MAL. Aggiungiamo una colonna booleana `has_synopsis` per distinguere i casi.

In [26]:
analyze(df_details['synopsis'])


  Nome serie                     synopsis
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               23,828  (82.29%)
  Null / NaN                     5,127  (17.71%)
  Stringhe vuote                 0
  Valori unici                   23,540  (81.30%)
  Valori duplicati               288

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  10  caratteri
  Lunghezza max                  3,750  caratteri
  Lunghezza media                377.40  caratteri
  Lunghezza mediana              251.0000  caratteri
  Dev. standard lunghezza        351.38
  IQR lunghezza                  499.0000

Valori di lunghezza estrema
───

**Osservazioni:**
- **5.127 valori nulli** (~17.7%): strutturali, anime privi di sinossi su MAL (spesso cortometraggi, CM, PV).
- **Aggiunta colonna `has_synopsis`**: booleano che indica la presenza di una sinossi.

In [27]:
df_details['has_synopsis'] = df_details['synopsis'].notna()
print(f'has_synopsis → True (con sinossi): {df_details["has_synopsis"].sum():,}  |  False: {(~df_details["has_synopsis"]).sum():,}')

has_synopsis → True (con sinossi): 23,828  |  False: 5,127


### 2.12 `rank`

Classifica globale MAL basata sullo score. I null sono strutturali: anime senza score sufficiente non ricevono un rank. Il dtype `float64` va convertito a `Int64` (nullable integer).

In [28]:
analyze(df_details['rank'])


  Nome serie                     rank
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               21,997  (75.97%)
  Null / NaN                     6,958  (24.03%)
  Zeri                           0  (0.00%)
  Positivi                       21,997   (75.97%)
  Negativi                       0   (0.00%)
  Valori unici                   17,765  (61.35%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.0000
  Max                            22,020.00
  Range                          22,019.00
  Media                          11,033.16
  Mediana                        11,008.00
  Moda/e                         2,651.00, 2,81

**Osservazioni:**
- **6.958 valori nulli** (~24.0%): strutturali, anime non classificati da MAL (score assente o insufficiente).
- **Conversione dtype**: da `float64` a `Int64` (nullable integer), dato che i valori sono interi.

In [29]:
df_details['rank'] = df_details['rank'].astype('Int64')
print(f'rank dtype   : {df_details["rank"].dtype}')
print(f'Null residui : {df_details["rank"].isna().sum():,}')

rank dtype   : Int64
Null residui : 6,958


### 2.13 `popularity`, `members`, `favorites`

Tre colonne numeriche che misurano l'interesse degli utenti MAL verso l'anime. Sono sempre valorizzate (anche gli anime senza voti hanno qualche utente nella lista). Il dtype `int64` è già corretto.

In [30]:
for col in ['popularity', 'members', 'favorites']:
    analyze(df_details[col])


  Nome serie                     popularity
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   21,954  (75.82%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            28,999
  Range                          28,998
  Media                          14,500.36
  Mediana                        14,497.00
  Moda/e                         10,785, 15,862, 21,379
 

**Osservazioni:**
- **Nessun valore nullo** in nessuna delle tre colonne.
- **Dtype corretto**: `int64` per tutte e tre.
- **Distribuzioni asimmetriche**: code lunghe verso destra, da tenere in considerazione nelle analisi.

Nessuna pulizia necessaria.

In [31]:
# Nessuna operazione richiesta
for col in ['popularity', 'members', 'favorites']:
    print(f'Null in {col:<12}: {df_details[col].isna().sum()}')

Null in popularity  : 0
Null in members     : 0
Null in favorites   : 0


### 2.14 `genres`, `studios`, `themes`, `demographics`, `producers`, `licensors`, `streaming`

Colonne contenenti liste serializzate come stringhe Python (es. `"['Comedy', 'Drama']"`).

In [32]:
list_cols = ['genres', 'studios', 'themes', 'demographics', 'producers', 'licensors', 'streaming']
for col in list_cols:
    analyze(df_details[col])


  Nome serie                     genres
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   931  (3.22%)
  Valori duplicati               28,024

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  90  caratteri
  Lunghezza media                16.96  caratteri
  Lunghezza mediana              13.0000  caratteri
  Dev. standard lunghezza        12.71
  IQR lunghezza                  14.0000

Valori di lunghezza estrema
──────────────────

**Osservazioni:**
- **Nessun valore nullo** in nessuna delle 7 colonne.
- **Liste vuote `[]` legittime**: anime senza studio, senza genere, senza streaming, ecc.
- **Formato stringa mantenuto**: sufficiente per le analisi testuali; deserializzabile con `ast.literal_eval` quando necessario.

Nessuna pulizia necessaria.

In [33]:
# Nessuna operazione richiesta
list_cols = ['genres', 'studios', 'themes', 'demographics', 'producers', 'licensors', 'streaming']
for col in list_cols:
    print(f'Null in {col:<14}: {df_details[col].isna().sum()}')

Null in genres        : 0
Null in studios       : 0
Null in themes        : 0
Null in demographics  : 0
Null in producers     : 0
Null in licensors     : 0
Null in streaming     : 0


### 2.15 `source`

Colonna categorica con il materiale di origine dell'anime (Manga, Novel, Original, …). I valori sono coerenti con la tassonomia MAL; non ci aspettiamo null.

In [34]:
analyze(df_details['source'])


  Nome serie                     source
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   17  (0.06%)
  Valori duplicati               28,938

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  4  caratteri
  Lunghezza max                  12  caratteri
  Lunghezza media                7.28  caratteri
  Lunghezza mediana              8.0000  caratteri
  Dev. standard lunghezza        2.11
  IQR lunghezza                  3.0000

Valori di lunghezza estrema
───────────────────────

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un valore di source.
- **17 valori distinti**: tutti coerenti con la tassonomia MAL, incluso `Unknown` (2.794 casi) che è un valore esplicito per anime con origine non documentata.

Nessuna pulizia necessaria.

In [35]:
# Nessuna operazione richiesta
print(f'Null in source : {df_details["source"].isna().sum()}')
print(df_details['source'].value_counts())

Null in source : 0
source
Original        12359
Manga            5508
Unknown          2794
Game             1473
Other            1326
Light novel      1199
Visual novel     1160
Novel             820
Web manga         654
4-koma manga      339
Picture book      284
Music             263
Mixed media       237
Book              224
Web novel         222
Card game          79
Radio              14
Name: count, dtype: int64


### 2.16 `rating`

Classificazione per età secondo la tassonomia MAL (G, PG, PG-13, R-17+, R+, Rx). I null sono strutturali: alcuni anime non hanno ancora una classificazione registrata.

In [36]:
analyze(df_details['rating'])


  Nome serie                     rating
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,355  (97.93%)
  Null / NaN                     600  (2.07%)
  Stringhe vuote                 0
  Valori unici                   6  (0.02%)
  Valori duplicati               28,349

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  11  caratteri
  Lunghezza max                  30  caratteri
  Lunghezza media                18.08  caratteri
  Lunghezza mediana              13.0000  caratteri
  Dev. standard lunghezza        6.67
  IQR lunghezza                  13.0000

Valori di lunghezza estrema
───────────────────

**Osservazioni:**
- **600 valori nulli** (~2.1%): strutturali, anime senza classificazione per età registrata su MAL.
- **6 valori distinti**: G, PG, PG-13, R-17+, R+, Rx, tutti coerenti con la classificazione MAL.

Nessuna pulizia necessaria.

In [37]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in rating : {df_details["rating"].isna().sum()}')
print(df_details['rating'].value_counts())

Null in rating : 600
rating
PG-13 - Teens 13 or older         10463
G - All Ages                       9066
PG - Children                      4421
R - 17+ (violence & profanity)     1593
Rx - Hentai                        1583
R+ - Mild Nudity                   1229
Name: count, dtype: int64


### 2.17 `episodes`

Numero di episodi dell'anime. I null sono strutturali: anime in corso o non ancora in onda possono non avere ancora un numero definitivo. Il dtype `float64` va convertito a `Int64` (nullable integer).

In [38]:
analyze(df_details['episodes'])


  Nome serie                     episodes
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,275  (97.65%)
  Null / NaN                     680  (2.35%)
  Zeri                           0  (0.00%)
  Positivi                       28,275   (97.65%)
  Negativi                       0   (0.00%)
  Valori unici                   261  (0.90%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.0000
  Max                            3,000.00
  Range                          2,999.00
  Media                          14.1301
  Mediana                        2.0000
  Moda/e                         1.0000
  Dev. standard  

**Osservazioni:**
- **680 valori nulli** (~2.3%): strutturali, anime in corso o non ancora in onda senza numero definitivo di episodi.
- **Conversione dtype**: da `float64` a `Int64` (nullable integer), dato che i valori sono interi.

In [39]:
df_details['episodes'] = df_details['episodes'].astype('Int64')
print(f'episodes dtype   : {df_details["episodes"].dtype}')
print(f'Null residui     : {df_details["episodes"].isna().sum()}')
print(f'Range            : {df_details["episodes"].min()} – {df_details["episodes"].max()}')

episodes dtype   : Int64
Null residui     : 680
Range            : 1 – 3000


### 2.18 `season`

Stagione di uscita dell'anime (spring, summer, fall, winter). I null sono strutturali: Movies, OVA, ONA, Music, CM e PV non appartengono a una stagione MAL.

In [40]:
analyze(df_details['season'])


  Nome serie                     season
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               6,266  (21.64%)
  Null / NaN                     22,689  (78.36%)
  Stringhe vuote                 0
  Valori unici                   4  (0.01%)
  Valori duplicati               6,262

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  4  caratteri
  Lunghezza max                  6  caratteri
  Lunghezza media                5.42  caratteri
  Lunghezza mediana              6.0000  caratteri
  Dev. standard lunghezza        0.91
  IQR lunghezza                  2.0000

Valori di lunghezza estrema
──────────────────────

**Osservazioni:**
- **22.689 valori nulli** (~78.4%): strutturali, Movies, OVA, ONA, Music, CM e PV non appartengono a una stagione MAL.
- **4 valori distinti**: `spring`, `summer`, `fall`, `winter`, tutti coerenti.

Nessuna pulizia necessaria.

In [41]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in season : {df_details["season"].isna().sum():,} ({df_details["season"].isna().mean()*100:.1f}%)')
print(df_details['season'].value_counts())

Null in season : 22,689 (78.4%)
season
spring    1972
fall      1825
winter    1302
summer    1167
Name: count, dtype: int64


### 2.19 `year`

Anno di uscita dell'anime. I null corrispondono esattamente a quelli di `season` (stessa logica strutturale). Il dtype `float64` va convertito a `Int64` (nullable integer).

In [42]:
analyze(df_details['year'])


  Nome serie                     year
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               6,266  (21.64%)
  Null / NaN                     22,689  (78.36%)
  Zeri                           0  (0.00%)
  Positivi                       6,266   (21.64%)
  Negativi                       0   (0.00%)
  Valori unici                   66  (0.23%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1,961.00
  Max                            2,026.00
  Range                          65.0000
  Media                          2,010.03
  Mediana                        2,014.00
  Moda/e                         2,023.00
  Dev. standar

**Osservazioni:**
- **22.689 valori nulli**: identici a quelli di `season` (stessa logica strutturale).
- **Conversione dtype**: da `float64` a `Int64` (nullable integer).

In [43]:
df_details['year'] = df_details['year'].astype('Int64')
print(f'year dtype   : {df_details["year"].dtype}')
print(f'Null residui : {df_details["year"].isna().sum():,}')
print(f'Range        : {df_details["year"].min()} – {df_details["year"].max()}')

year dtype   : Int64
Null residui : 22,689
Range        : 1961 – 2026


### 2.20 `explicit_genres`

Colonna destinata ai generi espliciti (es. Hentai). Verifichiamo se contiene informazione utile o se è interamente vuota.

In [44]:
analyze(df_details['explicit_genres'])


  Nome serie                     explicit_genres
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   1  (0.00%)
  Valori duplicati               28,954

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  2  caratteri
  Lunghezza media                2.00  caratteri
  Lunghezza mediana              2.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
────────────────

**Osservazioni:**
- La colonna contiene esclusivamente il valore `[]` (lista vuota) per tutte le 28.955 righe: **nessuna informazione utile**.
- **Rimozione della colonna**.

In [45]:
print(f'Valori unici in explicit_genres: {df_details["explicit_genres"].unique()}')
df_details.drop(columns=['explicit_genres'], inplace=True)
print('Colonna explicit_genres rimossa.')

Valori unici in explicit_genres: <StringArray>
['[]']
Length: 1, dtype: str
Colonna explicit_genres rimossa.


### 2.21 Chiave primaria `mal_id`

Verifichiamo che `mal_id` sia ancora univoco dopo tutte le operazioni di pulizia.

In [46]:
n_pk_dup = df_details.duplicated(subset=['mal_id'], keep=False).sum()
print(f'Duplicati su mal_id: {n_pk_dup}')

if n_pk_dup > 0:
    print('→ Rimozione duplicati sulla chiave primaria...')
    df_details.drop_duplicates(subset=['mal_id'], keep='first', inplace=True)
    print(f'Righe dopo la rimozione: {len(df_details):,}')
else:
    print('→ Chiave primaria già univoca, nessuna operazione richiesta.')

Duplicati su mal_id: 0
→ Chiave primaria già univoca, nessuna operazione richiesta.


## 3. Riepilogo e Salvataggio

Le operazioni di pulizia sono state effettuate colonna per colonna nella sezione 2. In questa sezione riepiloghiamo il risultato ed effettuiamo il salvataggio del dataset finale.

In [47]:
print('=== Riepilogo Dataset Pulito ===')
print(f'Righe originali      : {n_originale:>10,}')
print(f'Righe dopo cleaning  : {len(df_details):>10,}')
print(f'Righe rimosse totali : {n_originale - len(df_details):>10,}')
print()
print('Dtypes finali:')
print(df_details.dtypes)
print()
print('Valori mancanti residui:')
print(df_details.isnull().sum())

=== Riepilogo Dataset Pulito ===
Righe originali      :     28,955
Righe dopo cleaning  :     28,955
Righe rimosse totali :          0

Dtypes finali:
mal_id                          int64
title                             str
title_japanese                    str
url                               str
image_url                         str
type                              str
status                            str
score                         float64
scored_by                       Int64
start_date        datetime64[us, UTC]
end_date          datetime64[us, UTC]
synopsis                          str
rank                            Int64
popularity                      int64
members                         int64
favorites                       int64
genres                            str
studios                           str
themes                            str
demographics                      str
source                            str
rating                            str
episodes     

In [48]:
df_details.head()

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,source,rating,episodes,season,year,producers,licensors,streaming,has_image,has_synopsis
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,<NA>,2010-01-01 00:00:00+00:00,...,Original,G - All Ages,1,NaN,<NA>,['Nagoya Zokei University'],[],[],True,True
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503,2023-06-11 00:00:00+00:00,...,Original,PG-13 - Teens 13 or older,1,NaN,<NA>,[],[],[],True,True
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745,2007-01-18 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,"['Bandai Visual', 'CyberConnect2']",[],[],True,True
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373,2007-12-22 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,['Bandai Visual'],"['Funimation', 'Bandai Entertainment']",[],True,True
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317,2008-03-25 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,['Bandai Visual'],[],[],True,True


In [49]:
df_details.to_csv('../datasets_cleaned/details_clean.csv', index=False)
print('Salvato: datasets_cleaned/details_clean.csv')

Salvato: datasets_cleaned/details_clean.csv
